data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumerFactory.java

```diff
@@ -0,0 +1,98 @@
+    private final ShardManager shardManager;
+
+    private final Buffer<Record<Event>> buffer;
+
+    public ShardConsumerFactory(EnhancedSourceCoordinator enhancedSourceCoordinator, DynamoDbStreamsClient streamsClient, PluginMetrics pluginMetrics, ShardManager shardManager, Buffer<Record<Event>> buffer) {
+        this.streamsClient = streamsClient;
+        this.enhancedSourceCoordinator = enhancedSourceCoordinator;
+        this.pluginMetrics = pluginMetrics;
+        this.shardManager = shardManager;
+        this.buffer = buffer;
+
+    }
+
+    public Runnable createConsumer(StreamPartition streamPartition) {
+
+        LOG.debug("Try to create a thread for shard " + streamPartition.getShardId());
+
+        // Check and get the current state.
+        Optional<StreamProgressState> progressState = streamPartition.getProgressState();
+        String sequenceNumber = null;
+        Instant startTime = null;
+        boolean waitForExport = false;
+        if (progressState.isPresent()) {
+            sequenceNumber = progressState.get().getSequenceNumber();
+            waitForExport = progressState.get().shouldWaitForExport();
+            if (progressState.get().getStartTime() != 0) {
+                startTime = Instant.ofEpochMilli(progressState.get().getStartTime());
+            }
+        }
+
+        String shardIter = shardManager.getShardIterator(streamPartition.getStreamArn(), streamPartition.getShardId(), sequenceNumber);
+        if (shardIter == null) {
+            LOG.error("Unable to get a shard iterator, looks like the shard has expired");
+            return null;
+        }
+
+        StreamCheckpointer checkpointer = new StreamCheckpointer(enhancedSourceCoordinator, streamPartition);
+        String tableArn = getTableArn(streamPartition.getStreamArn());
+        TableInfo tableInfo = getTableInfo(tableArn);
+        StreamRecordConverter recordConverter = new StreamRecordConverter(buffer, tableInfo, pluginMetrics);
+
+        ShardConsumer shardConsumer = ShardConsumer.builder(streamsClient)
+                .recordConverter(recordConverter)
+                .checkpointer(checkpointer)
+                .shardIterator(shardIter)
+                .startTime(startTime)
+                .waitForExport(waitForExport)
+                .build();
+        return shardConsumer;
+    }
+
+    private TableInfo getTableInfo(String tableArn) {
+        GlobalState tableState = (GlobalState) enhancedSourceCoordinator.getPartition(tableArn).get();
+        TableInfo tableInfo = new TableInfo(tableArn, TableMetadata.fromMap(tableState.getProgressState().get()));
+        return tableInfo;
+    }
+
+    private String getTableArn(String streamArn) {
+        // e.g. Given a stream arn: arn:aws:dynamodb:us-west-2:xxx:table/test-table/stream/2023-07-31T04:59:58.190
+        // Returns arn:aws:dynamodb:us-west-2:xxx:table/test-table
+        return streamArn.substring(0, streamArn.lastIndexOf('/') - 7);
```

Comment:
This `7` is a magic number. I think it is for the `stream/`. It would be more readable to make this a constant:

```
private static final int STREAM_TO_TABLE_OFFSET = "stream/".length();
```